lib/internal/streams/readable.js

```diff
@@ -1284,9 +1322,265 @@ ObjectDefineProperties(ReadableState.prototype, {
+    // back to a minimal strategy that just specifies the highWaterMark
+    // and no size algorithm. Using a ByteLengthQueuingStrategy here
+    // is unnecessary.
+    let strategy;
+    if (objectMode) {
+      if (!CountQueuingStrategy) {
+        CountQueuingStrategy = require('internal/webstreams/queuingstrategies');
+      }
+      strategy = new CountQueuingStrategy({ highWaterMark });
+    } else {
+      strategy = { highWaterMark };
+    }
+
+    self
+      .on('error', function(err) {
+        const { controller } = this[kConsume];
+        controller.error(err);
+      })
+      .on('close', function() {
+        const { controller } = this[kConsume];
+        if (controller) {
+          controller.error(new AbortError());
+        }
+      });
+    const consume = self[kConsume] = {
+      type,
+      objectMode,
+      controller: null,
+      push(chunk) {
+        const { objectMode, controller } = this;
+
+        assert(controller);
+
+        if (chunk === null) {
+          // TODO: Do we need to wait for construct to finish?
+          controller.close();
+          this.controller = null;
+        } else {
+          if (!objectMode) {
+            if (typeof chunk === 'string') {
+              chunk = new Uint8Array(Buffer.from(chunk));
+            } else if (Buffer.isBuffer(chunk)) {
+              // Copy the Buffer to detach it from the pool.
+              chunk = new Uint8Array(chunk);
+            } else if (Stream._isUint8Array(chunk)) {
+              // Do nothing...
+            } else if (chunk != null) {
+              throw new ERR_INVALID_ARG_TYPE(
+                'chunk', ['string', 'Buffer', 'Uint8Array'], chunk);
+            }
+          }
+
+          // TODO: Does controller perform any type checks?
+          controller.enqueue(chunk);
+        }
+
+        return controller.desiredSize > 0;
+      },
+      stream: new ReadableStream({
+        async start(controller) {
+          consume.controller = controller;
+
+          const { _readableState: state } = self;
+
+          if (self[kReading]) {
+            while (controller.desiredSize > 0) {
+              const chunk = self.read();
+              if (chunk === null) {
+                break;
+              }
+              controller.enqueue(chunk);
+            }
+          } else {
+            const buffer = state.buffer;
+            while (buffer.length) {
+              controller.enqueue(buffer.shift());
+            }
+            state.length = 0;
+          }
+
+          if (!state.constructed) {
+            await EE.once(destroyImpl.kConstruct, self);
+          }
+
+          if (state.ended) {
+            controller.close();
+          }
+        },
+        pull() {
+          const n = consume.controller.desiredSize;
+
+          if (self[kReading]) {
+            self.read(n);
+          } else {
+            _read(self, n);
+          }
+        },
+        cancel(reason) {
+          self.destroy(reason);
```

Comment:
A bunch of these don't return the correct type (for example cancel should return a promise for "when the stream is canceled")